### PR TITLE
Map, add the marker, improve what have been added by #166

### PR DESCRIPTION
**Pull request by Qoookie**
# Summary

This PR introduces several improvements and new functionalities to the map feature, focusing on usability, clarity, and reliability.

# Task

- **Added marker placement:**
Users can now long-press on the map to place markers, enabling them to define and visualize organization areas.
Once at least three markers are set, an area can be created.

- **Improved error feedback:**
Error messages now indicate more precisely which step or component they originate from, simplifying debugging and user understanding.

- **Enhanced testing coverage:**
Added structured and comprehensive tests to improve reliability and ensure stable future development.

- **Added user tooltip:**
A short in-app tip guides new users through the map interaction process, helping them understand how to create areas.

_This PR description has been Generated by AI_
# Short Video
https://github.com/user-attachments/assets/d40992ad-cec6-497c-8397-9618a5818d33

### CODE REVIEW

File: app/src/main/java/com/android/sample/ui/map/MapScreen.kt
[the remark this answers is not in the document]
Comment (Qoookie) [reply app/src/main/java/com/android/sample/ui/map/MapScreen.kt]: done in [aeac269](https://github.com/swent-agendapp/agendapp/pull/197/commits/aeac2691bb09864d1b38b441964f3e02b81d507e)

File: app/src/androidTest/java/com/android/sample/ui/map/MapScreenTest.kt [hunk 1]
```diff
@@ -33,5 +34,32 @@ class MapScreenTest {
...
+
+    composeTestRule
+        .onNodeWithTag(MapScreenTestTags.TOOLTIP_BUTTON)
+        .assertIsDisplayed()
+        .performClick()
+    composeTestRule.onNodeWithTag(MapScreenTestTags.TOOLTIP_TEXT).assertIsDisplayed()
+  }
+
+  @Test
+  fun createAnAreaWith3Marker() {
```
Comment (Barde-E) [inline app/src/androidTest/java/com/android/sample/ui/map/MapScreenTest.kt]: This test should probably have another name to better reflect what it's testing :)

File: app/src/main/java/com/android/sample/ui/map/MapScreen.kt [hunk 2]
```diff
@@ -1,42 +1,96 @@
...
+import com.google.maps.android.compose.Marker
+import com.google.maps.android.compose.MarkerState
+import com.google.maps.android.compose.Polygon
 import com.google.maps.android.compose.rememberCameraPositionState
 import com.google.maps.android.ktx.MapsExperimentalFeature
 import kotlinx.coroutines.flow.distinctUntilChanged
 import kotlinx.coroutines.flow.filterNotNull
+import kotlinx.coroutines.launch
 
 object MapScreenTestTags {
```
Comment (HB616) [inline app/src/main/java/com/android/sample/ui/map/MapScreen.kt]: Good TestTag

File: app/src/main/java/com/android/sample/ui/map/MapViewModel.kt [hunk 3]
```diff
@@ -43,6 +61,74 @@ class MapViewModel(app: Application) : AndroidViewModel(app) {
...
+  fun createNewArea() {
+    _state.value.listNewMarker.forEach { marker -> mapRepository.addMarker(marker) }
+    try {
+      mapRepository.createArea(
+          label = _state.value.nextAreaName, markerIds = _state.value.listNewMarker.map { it.id })
+      _state.value = _state.value.copy(listNewMarker = emptyList(), nextAreaName = "Untitled")
+      fetchAllArea()
+    } catch (e: IllegalArgumentException) {
+      _state.value = _state.value.copy(errorMessage = e.message)
+    }
```
Comment (Barde-E) [inline app/src/main/java/com/android/sample/ui/map/MapViewModel.kt]: This method should not allow an empty `nextAreaName` or even a `listNewMarker` containing less than 3 items (and if it's the case, it could be good to let the User know, by setting for example the errorMessage accordingly)

One way of doing it could be something like this :
```Kotlin
fun createNewArea() {
    if (_state.value.nextAreaName.isBlank()) {
        _state.update { it.copy(errorMessage = "Area name cannot be empty") }
        return
    }

    if (_state.value.listNewMarker.size < 3) {
        _state.update { it.copy(errorMessage = "You need at least 3 markers to create an area") }
        return
    }

    // your creation logic as it is right now
}
```
If you choose to follow this idea of `errorMessage`, think to update the `string.xml` to accordingly :)

File: app/src/test/java/com/android/sample/ui/map/MapViewModelTest.kt
Comment (HB616) [inline app/src/test/java/com/android/sample/ui/map/MapViewModelTest.kt]: Nice tests to test the map VM features.

File: app/src/main/java/com/android/sample/ui/map/MapViewModel.kt [hunk 4]
```diff
@@ -43,6 +61,74 @@ class MapViewModel(app: Application) : AndroidViewModel(app) {
...
+   *
+   * @param coordinate The geographical position (latitude and longitude) of the new marker.
+   */
+  fun addNewMarker(coordinate: LatLng) {
+    _state.value =
+        _state.value.copy(
+            listNewMarker =
+                listOf(Marker(location = Location(coordinate.latitude, coordinate.longitude))) +
+                    _state.value.listNewMarker)
+  }
```
Comment (HB616) [inline app/src/main/java/com/android/sample/ui/map/MapViewModel.kt]: The order is added from back to front.

**Review by HB616 — CHANGES_REQUESTED** [review]
The code looks solid overall with no major issues. I’ve added a few minor suggestions in the comments — feel free to adjust them as you see fit.

File: app/src/main/java/com/android/sample/ui/map/MapViewModel.kt (hunk 4 above)
Comment (Qoookie) [reply app/src/main/java/com/android/sample/ui/map/MapViewModel.kt]: Drawing the shape clockwise or counterclockwise produces the same result, so this shouldn’t cause any issue.

File: app/src/main/java/com/android/sample/ui/map/MapScreen.kt [hunk 5]
```diff
@@ -84,17 +151,135 @@ fun MapScreen(
                         contentDescription = "Back")
                   }
             },
-        )
+            actions = {
+              TooltipBox(
```
Comment (Barde-E) [inline app/src/main/java/com/android/sample/ui/map/MapScreen.kt]: I really like the rich `TooltipBox` here ! It provides a detailed guide on how to use this screen, which will be super helpful for first-time users discovering this feature !

File: app/src/main/java/com/android/sample/ui/map/MapViewModel.kt [hunk 6]
```diff
@@ -43,6 +61,74 @@ class MapViewModel(app: Application) : AndroidViewModel(app) {
...
+      _state.value = _state.value.copy(errorMessage = e.message)
+    }
+  }
+
+  /**
+   * Add a new marker to the temporary list of markers used for creating a new area.
+   *
+   * @param coordinate The geographical position (latitude and longitude) of the new marker.
+   */
+  fun addNewMarker(coordinate: LatLng) {
```
[the remark this answers is not in the document]
Comment (Qoookie) [reply app/src/main/java/com/android/sample/ui/map/MapViewModel.kt]: done in [5351f0d](https://github.com/swent-agendapp/agendapp/pull/197/commits/5351f0dd99973e0bcd2828185717c8d038c7989c)

File: app/src/main/java/com/android/sample/ui/map/MapViewModel.kt [hunk 7]
```diff
@@ -43,6 +61,74 @@ class MapViewModel(app: Application) : AndroidViewModel(app) {
...
+   * Create a new area based on the markers currently stored in the temporary list.
+   *
+   * Each marker is first saved in the repository. Then, a new area is created using the current
+   * `nextAreaName`. After successful creation, the list of new markers and the area name are reset,
+   * and all areas are re-fetched to refresh the state.
+   *
+   * If an invalid area name or configuration is detected, the function catches
+   * [IllegalArgumentException] and updates the state with the corresponding error message.
+   */
+  fun createNewArea() {
```
[the remark this answers is not in the document]
Comment (Qoookie) [reply app/src/main/java/com/android/sample/ui/map/MapViewModel.kt]: done in [5351f0d](https://github.com/swent-agendapp/agendapp/pull/197/commits/5351f0dd99973e0bcd2828185717c8d038c7989c)

File: app/src/main/java/com/android/sample/ui/map/MapScreen.kt [hunk 8]
```diff
@@ -84,17 +151,135 @@ fun MapScreen(
...
-            modifier =
-                Modifier.fillMaxSize()
-                    .testTag(MapScreenTestTags.GOOGLE_MAP_SCREEN)
-                    .padding(padding),
-            cameraPositionState = cameraPositionState,
-            properties =
-                MapProperties(
-                    isMyLocationEnabled = uiState.hasPermission, mapType = MapType.NORMAL)) {}
+        Column(modifier = Modifier.fillMaxSize()) {
+          Box(modifier = Modifier.weight(1f).fillMaxWidth().padding(padding)) {
```
[the remark this answers is not in the document]
Comment (Qoookie) [reply app/src/main/java/com/android/sample/ui/map/MapScreen.kt]: done in [5351f0d](https://github.com/swent-agendapp/agendapp/pull/197/commits/5351f0dd99973e0bcd2828185717c8d038c7989c)

File: app/src/main/java/com/android/sample/ui/map/MapViewModel.kt (hunk 3 above)
Comment (Qoookie) [reply app/src/main/java/com/android/sample/ui/map/MapViewModel.kt]: I don’t think the blank value is an issue since there’s already a default value, and the area name isn’t really used at this stage.
Regarding the three markers, the message is already displayed because the createArea method performs this check, so to avoid redundant validation, I chose not to check it here again.

File: app/src/test/java/com/android/sample/ui/map/MapViewModelTest.kt [hunk 9]
```diff
@@ -13,7 +18,70 @@ class MapViewModelTest {
...
+
+    assertEquals(3, vm.state.value.listNewMarker.size)
+
+    vm.deleteMarker(vm.state.value.listNewMarker[0].id)
+
+    assertEquals(2, vm.state.value.listNewMarker.size)
+  }
+
+  @Test
+  fun `create area with 2 marker make an error and clean it`() {
```
[the remark this answers is not in the document]
Comment (Qoookie) [reply app/src/test/java/com/android/sample/ui/map/MapViewModelTest.kt]: The only edge case here is when the marker count is 2. The program behaves the same for two markers or fewer, so adding a separate check would just duplicate code.

**Review by HB616 — APPROVED** [review]
The code is in excellent shape, and all previously raised issues have been fully addressed. The implementation aligns well with the intended UX and significantly improves the overall map feature.
